### ISSUE TITLE
[clang-format] Consistent style options for lists

### ISSUE THREAD

**Issue by Stadly** [llvm/llvm-project#62676]
There is a lack of consistency when it comes to the way lists can be formatted by Clang-Format. This applies both to how the styles are configured and which styles are supported.

For example, parameter lists can be bin-packed with the style option `BinPackParameters`, member initializer lists with `PackConstructorInitializers: BinPack`, while inheritance lists cannot be bin-packed.

I envision full consistency for styling lists, so options for alignment, breaking, indentation, and packing can be set consistently for all types of lists. This would simplify both the implementation, documentation and configuration, and make Clang-Format both more flexible and easier to maintain.

My thinking is to create nested configuration flags for each list type. Here is an example for member initializer lists (could also be called constructor initializer lists if we want to):
```
MemberInitializerList:
  Align: boolean
  Break: BeforeColon/BeforeComma/AfterColon/AfterComma
  IndentWidth: unsigned
  Pack: Never/BinPack/CurrentLine/NextLine/NextLineOnly
```

- `Align` provides alignment similar to `AlignAfterOpenBracket: Align` for functions.
- `Break` replaces `BreakConstructorInitializers`
- `IndentWidth` replaces `ConstructorInitializerIndentWidth`
- `Pack` replaces `PackConstructorInitializers`, with the additional option `AfterComma`

We would have similar nested configuration flags for the other lists types. For example parameter lists:
```
ParameterList:
  Align: boolean
  Break: BeforeParens/BeforeComma/AfterParens/AfterComma
  IndentWidth: unsigned
  Pack: Never/BinPack/CurrentLine/NextLine/NextLineOnly
```

- `Align` and `Break` together replaces `AlignAfterOpenBracket`, but provides greater flexibility
- `Pack: BinPack` replaces `BinPackParameters`
- `Pack: NextLine` replaces `AllowAllParametersOfDeclarationOnNextLine`

We should support at least these different types of lists:
- Argument lists
- Enumerator lists
- Inheritance lists
- Initializer lists
- Member initializer lists
- Parameter lists

This change would lead to improvements for:
- Documentation and configuration
  - Get rid of inconsistencies like
    - `ConstructorInitializerIndentWidth` also applies to inheritance lists
    - `BinPackParameters: true` for parameter lists, but `PackConstructorInitializers: BinPack` for member initializer lists
    - `BreakInheritanceList` has option `AfterComma`, while `BreakConstructorInitializers` doesn't
    - `OfDeclaration` is part of the name `AllowAllParametersOfDeclarationOnNextLine` but not `BinPackParameters`
  - No more searching around for the style options applicable for a list type; all the options are grouped as nested configurations
  - No more wondering which style options are supported for a list type; they all support the same set of options (with tiny variations like `BeforeColon` vs `BeforeParens`)
- Implementation
  - The implementation can be generic and not need separate handling for each list type. By looking at the context it is in, the appropriate set of style options can be loaded, and then the algorithm can be agnostic to which type of list it is working on.
- Maintenance and future enhancements
  - Adding a new feature to one of the list types would add support for all list types
  - Fixing a bug in one of the list types would fix it for all list types

The change would deprecate the following style options:
- `AlignAfterOpenBracket`
- `AllowAllArgumentsOnNextLine`
- `AllowAllParametersOfDeclarationOnNextLine`
- `AllowShortEnumsOnASingleLine`
- `BinPackArguments`
- `BinPackParameters`
- `BracedInitializerIndentWidth`
- `BreakConstructorInitializers`
- `BreakInheritanceList`
- `ConstructorInitializerIndentWidth`
- `PackConstructorInitializers`

It might also be possible to deprecate these options:
- `BreakArrays`
- `ObjCBinPackProtocolList`

We might have to update `ExperimentalAutoDetectBinPacking` to work with the new style options.

What do you think?

**Comment by owenca** [comment]
Sounds great, at least in theory. Tagging @mydeveloperday @HazardyKnusperkeks @rymiel.

**Comment by rymiel** [comment]
Yes, it is nice in theory, but I'm not sure about practice, especially something like this:

> The implementation can be generic and not need separate handling for each list type

It's not like all these different options exist on purpose in the first place

**Comment by HazardyKnusperkeks** [comment]
I'd say give it a go, and we will review it.

**Comment by Stadly** [comment]
Thanks, then I'll give it a go :)

I don't have any prior experience with the Clang-Format code base. If you have any immediate thoughts about things that may prove challenging or things that I should keep in mind, that would be highly appreciated.

If I run into any issues or questions, or if I need to discuss different approaches, would this issue be the right place to raise them?

**Comment by mskripnik** [comment]
Handling lists seems to be really the only weak point of clang-format so far (at least for me). Just check how initializer_lists are formatted:
![image](https://github.com/llvm/llvm-project/assets/156424199/3bb0ba8b-6f30-4b62-9d52-5dc2fa575d44)
That looks just awful!

**Comment by nmschulte** [comment]
Any progress updates @Stadly?

**Comment by Stadly** [comment]
No, sorry. I changed jobs. No longer working with c++
